### PR TITLE
fix: #101 Payload type consistency is not enforced.

### PR DESCRIPTION
**Pull request by gvelo**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/payload_storage/simple_payload_storage.rs [hunk 1]
```diff
@@ -31,53 +33,52 @@ impl SimplePayloadStorage {
...
-                } else {
-                    Err(OperationError::TypeError {
-                        field_name: key.to_owned(),
-                        expected_type: format!("{:?}", schema_type),
-                    })
-                }
-            }
-        };
-    }
+    // fn update_schema_value(
```
Comment (agourlay) [inline lib/segment/src/payload_storage/simple_payload_storage.rs:53]: Can we delete this commented code?

Comment (gvelo) [reply lib/segment/src/payload_storage/simple_payload_storage.rs]: done

File: lib/segment/src/payload_storage/schema_storage.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            schema: RwLock::new(schema),
+        }
+    }
+
+    pub fn update_schema_value(
+        &self,
+        key: PayloadKeyTypeRef,
+        value: &PayloadType,
+    ) -> OperationResult<()> {
+        let mut schema = self.schema.write();
```
[the remark this answers is not in the document]
Comment (gvelo) [reply lib/segment/src/payload_storage/schema_storage.rs]: updated, now the exclusive write lock only protects the update branch.

File: lib/segment/src/payload_storage/schema_storage.rs [hunk 3]
```diff
@@ -21,9 +21,10 @@ impl SchemaStorage {
         key: PayloadKeyTypeRef,
         value: &PayloadType,
     ) -> OperationResult<()> {
-        let mut schema = self.schema.write();
+        let schema = self.schema.upgradable_read();
```
Comment (generall) [inline lib/segment/src/payload_storage/schema_storage.rs:24]: `upgradable_read` unfortunately is still exclusive for other `upgradable_read`, so this change does not help completely.
I would add another quick check with just a `read` lock before that

Comment (gvelo) [reply lib/segment/src/payload_storage/schema_storage.rs]: refactored to use a Double-checked locking on schema update.

**Review by generall — APPROVED** [review]

**Review by agourlay — APPROVED** [review]

**Pull request merged**